platform/pulp/plugin/changeset/main.py

```diff
@@ -0,0 +1,200 @@
+        log.info(activity)
+        with Batch(DownloadIterator(self.adds, self.deferred)) as batch:
+            with ProgressBar(message=activity, total=len(self.adds)) as bar:
+                for plan in batch():
+                    artifact = plan.download.attachment
+                    artifact.model.downloaded = True
+                    content = artifact.content
+                    if not plan.succeeded:
+                        task = Task()
+                        task.append_non_fatal_error(plan.error)
+                        bar.increment()
+                        report = ChangeReport(ChangeReport.ADDED, content.model)
+                        report.errors.append(plan.error)
+                        yield report
+                        continue
+                    content.settle()
+                    if not content.settled:
+                        continue
+                    bar.increment()
+                    with transaction.atomic():
+                        self._update_catalog(content)
+                        self._associate(content.model)
+                    report = ChangeReport(ChangeReport.ADDED, content.model)
+                    yield report
+
+    def _update_catalog(self, content):
+        """
+        Update the content catalog.
+        Add (or update) an entry for each artifact.
+
+        Args:
+            content (RemoteContent): A newly added content.
+        """
+        for artifact in content.artifacts:
+            log.info(_('Cataloging: {a}').format(a=self))
+            try:
+                entry = DownloadCatalog()
+                entry.importer = self.importer
+                entry.url = artifact.download.url
+                entry.artifact = artifact.model
+                entry.save()
+            except IntegrityError:
+                entry = DownloadCatalog.objects.filter(
+                    importer=self.importer,
+                    artifact=artifact)
+                entry.url = artifact.download.url
+                entry.save()
+
+    def _remove_content(self):
+        """
+        Delete unwanted content.
+
+        Yields:
+            DeleteReport: A report for each removed content.
+        """
+        activity = _('Removing content.')
+        log.info(activity)
+
+        with ProgressBar(message=activity, total=len(self.deletes)) as bar:
+            for content in self.deletes:
+                self._disassociate(content)
+                bar.increment()
+                report = ChangeReport(ChangeReport.DELETED, content)
+                yield report
+
+    def apply(self):
+        """
+        Apply the changes to the repository based on download policy.
+
+        Yields:
+            ChangeReport: For each change applied.
+        """
+        log.info(_('Apply'))
+        for report in self._add_content():
+            yield report
+        for report in self._remove_content():
+            yield report
+
+
+class SizedIterable(Sized, Iterable):
```

Comment:
Even though this class is not an important part of the PR, I have some concerns about it, I am not sure if you agree or not :)

If I understand correctly, this class is needed so we know the number of items for the unsized iterables like generators  because in case of sequence types which are sized `__len__` is defined. 
I think it is hard to support properly `len` for unsized by design iterable.
We are claiming that we support protocol both for `Sized` and for `Iterable` aka guarantee that number of iterations will be equal to what `__len__` returns. At the same time `length` is defined once in `__init__` and `__len__` returns always this number but `self.iterable` can be changed independently or just incorrectly initialized.
I see that `SizedIterable` is currently used for progress only and in the worst case the counters will be wrong but this class is also meant to be used by plugin writers. There could be various bugs which is hard to find in case iterable and length are different even just by mistake.

If you think this class is really needed, I suggest at least to improve it to avoid honest/simple mistakes.
Maybe make it immutable to some extent (or  less accessible like `self.__iterable`) and keep track of actual number of iterations and throw some error in case of discrepancy with `self.length`?

What do you think?